### PR TITLE
ENH: map tile overlays

### PR DESCRIPTION
**Pull request by jpn--**
This addresses #89.  Tile sets identified in providers that are known to be overlay tiles, with an embedded alpha channel, using the [same named rules from leaflet](https://github.com/leaflet-extras/leaflet-providers/blob/9eb968f8442ea492626c9c8f0dac8ede484e6905/preview/preview.js#L56-L70)) or that have defined opacity less than 1.0 are drawn over other data (with a default zorder of 9, unless this is overridden manually).  This will typically put the overlay tiles on top of other baseman tiles (almost certainly always good) and also on top of other data being plotted by the user.  This second outcome is sometimes good, sometimes not, and I'm not sure if this default setting is reasonable.  But, even setting zorder to 1 will put the labels on top of other user content unless the user intentionally put a zorder value on what they draw.  I am certainly open to implementing other options here if anyone cares deeply (or shallowly).

One side effect of this change is that all map tiles are processed with the alpha channel intact.  This will add a bit to memory requirements, but likely not enough for anyone to notice (other than the tests, which I've edited to accommodate the alpha channels across the board.)

### CODE REVIEW

**Comment by jpn--** [comment]
Having the default zorder as noted above lets you do this, putting labels over both the background and your data, without worrying about the order of commands to put it together:

image

**Comment by JSybrandt** [comment]
Any idea if/when a release is going to get this feature?

**Comment by darribas** [comment]
Thanks for this @jpn-- !!! I haven't had a second to review it myself, but does not look to onerous. Right now, our efforts are centered on settling the providers api, which we already have decided what to do (#90) and with that I'd be keen to push `1.0` this spring. If this is not too onerous and we can get it reviewed, I think it'd be a very cool feature to have!

**Comment by jpn--** [comment]
I noticed development here recently kicked into "active", and that some of the recent changes wouldn't merge cleanly with the changes I made to the tests for the overlay feature, so I went ahead and tweaked my changes so that they do merge cleanly, hopefully this makes it easier to review this now.  Thanks to you all for making this awesome tool!

File: contextily/plotting.py [hunk 1]
```diff
@@ -222,6 +226,44 @@ def _reproj_bb(left, right, bottom, top, s_crs, t_crs):
...
+        'OpenPtMap',
+        'OpenRailwayMap',
+        'OpenFireMap',
+        'SafeCast'
+    ]
+    import re
+    return bool(re.match('(' + '|'.join(overlayPatterns) + ')', url.get('name','')))
+
+
+# if (layer.options.opacity && layer.options.opacity < 1) {
```
Comment (jorisvandenbossche) [inline contextily/plotting.py:249]: Can you add a comment and link to the original source to explain this part?

File: tests/test_ctx.py [hunk 2]
```diff
@@ -361,9 +371,42 @@ def test_add_basemap():
...
     assert ax.get_ylim() == (y1, y2)
-    assert ax.images[0].get_array().sum() == 464751694
-    assert ax.images[0].get_array().shape == (980, 862, 3)
-    assert_array_almost_equal(ax.images[0].get_array().mean(), 183.38608756727749)
+
+    assert ax.images[0].get_array()[:,:,:3].sum() == 464536503
+    assert ax.images[0].get_array().shape == (980, 862, 4)
+    assert_array_almost_equal(ax.images[0].get_array()[:,:,:3].mean(), 183.301175)
+
+	#<<<<<<< HEAD
```
Comment (jorisvandenbossche) [inline tests/test_ctx.py:379]: some leftover from merging in master?

File: tests/test_ctx.py [hunk 3]
```diff
@@ -361,9 +371,42 @@ def test_add_basemap():
...
+    assert_array_almost_equal(ax.images[0].get_array().mean(), 200.939189)
+
+    x1, x2, y1, y2 = [
+        -11740727.544603072,
+        -11701591.786121061,
+        4852834.0517692715,
+        4891969.810251278,
+    ]
+
+    # Test web basemap with overlay layer
```
Comment (jorisvandenbossche) [inline tests/test_ctx.py:390]: Can you start a new test function for this? (eg `def test_add_basemap_overlay`, the existing one is already a bit too long)

File: tests/test_ctx.py [hunk 4]
```diff
@@ -361,9 +371,42 @@ def test_add_basemap():
...
+    assert ax.axis() == ax_extent
+
+    assert ax.images[1].get_array().sum() == 1653387
+    assert ax.images[1].get_array().shape == (256, 256, 4)
+    assert_array_almost_equal(ax.images[1].get_array().mean(), 6.3071708679)
+	#=======
+    #assert ax.images[0].get_array()[:,:,:3].sum() == 464751694
+    #assert ax.images[0].get_array().shape == (980, 862, 4)
+    #assert_array_almost_equal(ax.images[0].get_array().mean(), 183.38608756727749)
+	#>>>>>>> be7ab8b35dd3fed262964f28317a2da5f232809f
```
Comment (jorisvandenbossche) [inline tests/test_ctx.py:409]: same here, some merge leftovers

File: tests/test_ctx.py (hunk 3 above)
Comment (jpn--) [reply tests/test_ctx.py]: Test is broken out. Also fixed the test so it points to Boulder (where there are labels) and not the Atlantic Ocean (where there's nothing).

File: contextily/plotting.py (hunk 1 above)
Comment (jpn--) [reply contextily/plotting.py]: Link back to the original leaflet.js version in docstring for this function.

**Comment by jpn--** [comment]
> We probably want to add some examples for this in the notebook about `add_basemap`, once that exists (so don't worry about that for now)

Glad to contribute!  When the rest of a notebook about `add_basemap` exists, if you want me to circle back and put in an example, tag me and I can help with that too.

**Comment by jpn--** [comment]
I needed to use this so I updated it for myself.  I also added an example in the add_basemap notebook.

**Comment by darribas** [comment]
This looks fantastic! @jorisvandenbossche do the changes look good to you? If so, we can merge, and I think it'd be good to push it as 1.1, it's a really cool little feature! (I'm happy to take care of the release)

**Comment by danielsclint** [comment]
This feature is really cool and nice to have. I stumbled on a need for this, and @jpn-- patch got me fixed up right away.

**Comment by darribas** [comment]
Hello everyone, sorry for my slow moving on this (my bandwidth is what it is...). Just to say I'm planning on having a look at this PR in the coming days and if it's all good, happy to merge and have this on contextily. Once merged, I think this is good enough to cut a 1.1 release so folks can enjoy it more widely. Hopefully more on this soon...

**Comment by jpn--** [comment]
Sorry @darribas, my experiment with commit hooks was too clever by half, or not clever enough? In any case, the notebook should be all better now.

**Review by darribas — APPROVED** [review]
This is fantastic!!! I've had a more in detail look to get my hear around and this is so simple yet so powerful! If you consider this good to go, I'm very happy to merge it, just OK this comment

**Comment by jpn--** [comment]
I'm not clear if "you" is me or @jorisvandenbossche, I'm happy to see this merged either way.

**Comment by darribas** [comment]
Just wanted to make sure there wasn't any extra add to the PR. THANK you very much for the contribution this is fantastic! Hopefully I can cut a quick release in the coming days so it's available on pip/conda to everyone!

**Pull request merged**